### PR TITLE
distsql: avoid ReadAll allocation.

### PR DESCRIPTION
**Pull request by coocood**
Use a specific ReadCloser can get the underlying bytes directly.

The BytesPool can be used to reuse large bytes later, currently bytes is not reused.

### CODE REVIEW

**Comment by coocood** [comment]
Memstats after a query which index double read 1 million rows:

Master:
```
# runtime.MemStats
# Alloc = 1985803616
# TotalAlloc = 4365147096
```

This branch:
```
# runtime.MemStats
# Alloc = 1303384504
# TotalAlloc = 1676585832
```

**Comment by ngaut** [comment]
Cooool

File: distsql/distsql.go [hunk 1]
```diff
@@ -170,12 +171,16 @@ type partialResult struct {
...
 	pr.resp = new(tipb.SelectResponse)
-
-	b, err := ioutil.ReadAll(pr.reader)
-	pr.reader.Close()
-	if err != nil {
-		pr.done <- errors.Trace(err)
-		return
+	var b []byte
+	var err error
+	if bpReader, ok := pr.reader.(*bytespool.ReadCloser); ok {
```
Comment (ngaut) [inline distsql/distsql.go]: s/bpReader/rc/

File: util/bytespool/bytespool.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	closed bool
+}
+
+// Close implements the io.ReadCloser interface.
+// It frees the origin bytes allocated from the pool
+func (r *ReadCloser) Close() error {
+	if r.closed {
+		return nil
+	}
+	r.pool.Free(r.origin)
```
Comment (shenli) [inline util/bytespool/bytespool.go]: Seems pool is only use here in ReadCloser. When will we use it to allocate bytes slice?

Comment (coocood) [inline util/bytespool/bytespool.go]: When coprocessor refactor is finished.

File: util/bytespool/bytespool_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+func TestT(t *testing.T) {
+	TestingT(t)
+}
+
+var _ = Suite(&testBytesPoolSuite{})
+
+type testBytesPoolSuite struct{}
+
+func (s *testBytesPoolSuite) TestBytesPool(c *C) {
+	testCases := []struct {
```
Comment (ngaut) [inline util/bytespool/bytespool_test.go]: s/testCases/bptests/
Please ref https://github.com/golang/go/wiki/TableDrivenTests

**Comment by tiancaiamao** [comment]
I doubt this change. it brings much complexity but gains less.
What it does is just move something (already done) from runtime to user space.
And it optimize allocation here, while memory footprint really matters currently.

File: distsql/distsql.go [hunk 4]
```diff
@@ -170,12 +171,16 @@ type partialResult struct {
...
-	b, err := ioutil.ReadAll(pr.reader)
-	pr.reader.Close()
-	if err != nil {
-		pr.done <- errors.Trace(err)
-		return
+	var b []byte
+	var err error
+	if rc, ok := pr.reader.(*bytespool.ReadCloser); ok {
+		b = rc.SharedBytes()
+	} else {
```
Comment (ngaut) [inline distsql/distsql.go]: The else should be removed, it should never happen.

**Comment by ngaut** [comment]
There are lots of unrelated functions. I prefer to remove them and make the interface clean.

File: distsql/distsql.go (hunk 4 above)
Comment (coocood) [inline distsql/distsql.go]: It happens in test.

**Comment by coocood** [comment]
@ngaut PTAL
Added TODO to change the ReadCloser interface.
When we supported gRPC stream, we need a better interface.

**Comment by shenli** [comment]
LGTM

File: util/bytespool/bytespool.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
+package bytespool
```
Comment (zimulala) [inline util/bytespool/bytespool.go]: Add license.

File: util/bytespool/bytespool.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+// Alloc allocates a bytes which has the size of power of two.
+// The caller should keep the origin bytes and use the returned data.
+// When finished using, the origin bytes should be freed to the pool.
+// The allocated data may not have zero value.
+func (bp *BytesPool) Alloc(size int) (origin, data []byte) {
+	if size > maxSize {
+		return nil, make([]byte, size)
+	}
+	var i uint
+	for size > (baseSize << i) {
```
Comment (BusyJay) [inline util/bytespool/bytespool.go]: Is there any function like `leading_zeros` in go? So we don't need to loop here.

**Comment by coocood** [comment]
@BusyJay @zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**